src/java.base/windows/classes/sun/nio/ch/WindowsAsynchronousFileChannelImpl.java

```diff
@@ -446,13 +448,15 @@ public void run() {
             // Substitute a native buffer if not direct
             if (dst instanceof DirectBuffer) {
                 buf = dst;
-                address = ((DirectBuffer)dst).address() + pos;
+                address = NIO_ACCESS.getBufferAddress(dst) + pos;
             } else {
+                // buf can never be backed by a memory segment
                 buf = Util.getTemporaryDirectBuffer(rem);
-                address = ((DirectBuffer)buf).address();
+                address = NIO_ACCESS.getBufferAddress(buf);
             }
 
             boolean pending = false;
+            NIO_ACCESS.acquireSession(buf);
```

Comment:
Here, we acquire the session *after* we have obtained the address. This is safe as we do not touch the segment before it is acquired. If a segment is deallocated before we try to acquire the session, an exception will be thrown.

Reply:
Is there documentation on when sessions should be acquired/released? Is this only for when using MemorySegment? In security area, the bytes are passed to the JNI code which calls the native library to process the bytes and I wonder if sessions should be acquired/released for these.

Reply:
Never mind, I see that there are already acquireSession/releaseSession calls when the ByteBuffer objects are accessed.